spanner/benchmark/ycsb.py

```diff
@@ -0,0 +1,258 @@
+
+
+def open_database(parameters):
+  """Opens a database specified by the parameters from parse_options()."""
+  spanner_client = spanner.Client()
+  instance_id = parameters['cloudspanner.instance']
+  instance = spanner_client.instance(instance_id)
+  database_id = parameters['cloudspanner.database']
+  pool = spanner.BurstyPool(int(parameters['num_worker']))
+  database = instance.database(database_id, pool=pool)
+
+  return database
+
+
+def load_keys(database, parameters):
+  """Loads keys from database."""
+  del KEYS[:]
+  results = database.execute_sql('SELECT u.id FROM %s u' % parameters['table'])
+
+  for row in results:
+    KEYS.append(row[0])
+
+
+def read(database, table, key):
+  """Does a single read operation."""
+  with database.snapshot() as snapshot:
+    result = snapshot.execute_sql('SELECT u.* FROM %s u WHERE u.id="%s"' %
+                                  (table, key))
+    for row in result:
+      key = row[0]
+      for i in range(10):
+        field = row[i + 1]
+
+
+def update(database, table, key):
+  """Does a single update operation."""
+  field = random.randrange(10)
+  value = ''.join(random.choice(string.printable) for i in range(100))
+  with database.batch() as batch:
+    batch.update(table=table, columns=('id', 'field%d' % field),
+                 values=[(key, value)])
+
+
+def insert(database, table, key):
+  """Does a single insert operation."""
+  raise NotImplementedError('Insert is not implemented.')
+
+
+def scan(database, table, key):
+  """Does a single scan operation."""
+  raise NotImplementedError('Scan is not implemented.')
+
+
+def DoOperation(database, table, operation, latencies_ms):
+  """Does a single operation and records latency."""
+  key = random.choice(KEYS)
+  start = timeit.default_timer()
+  if operation == 'read':
+    read(database, table, key)
+  elif operation == 'update':
+    update(database, table,  key)
+  elif operation == 'insert':
+    insert(database, table, key)
+  elif operation == 'scan':
+    scan(database, table, key)
+  else:
+    raise Exception('Unknown operation: %s' % operation)
+  end = timeit.default_timer()
+  latencies_ms[operation].append((end - start) * 1000)
+
+
+def aggregate_metrics(latencies_ms, duration_ms, num_bucket):
+  """Aggregates metrics."""
+  overall_op_count = 0
+  op_counts = {}
+  for operation in latencies_ms.keys():
+    op_counts[operation] = len(latencies_ms[operation])
+    overall_op_count += op_counts[operation]
+
+  print '[OVERALL], RunTime(ms), %f' % duration_ms
+  print '[OVERALL], Throughput(ops/sec), %f' % (float(overall_op_count) /
+                                                duration_ms * 1000.0)
+
+  for operation in op_counts.keys():
+    operation_upper = operation.upper()
+    print '[%s], Operations, %d' % (operation_upper, op_counts[operation])
+    print '[%s], AverageLatency(us), %f' % (
+        operation_upper, numpy.average(latencies_ms[operation]) * 1000.0)
+    print '[%s], LatencyVariance(us), %f' % (
```

Comment:
Latency Deviation?  Variance is unclear if you meant np,var

Reply:
Actually I think it means np.var. changed.

Reply:
If it's variance, maybe the units need to be changed also.

Reply:
This follows exactly the Java YCSB, which says LatencyVariance and use (us) as the unit. Since we are modeling this following Java YCSB, I think probably we can just use the same?